Pcap++/src/PcapLiveDevice.cpp

```diff
@@ -488,41 +505,68 @@ int PcapLiveDevice::startCaptureBlockingMode(OnPacketArrivesStopBlocking onPacke
 	m_cbOnPacketArrivesBlockingMode = onPacketArrives;
 	m_cbOnPacketArrivesBlockingModeUserCookie = userCookie;
 
-	long startTimeSec = 0, startTimeNSec = 0;
-	clockGetTime(startTimeSec, startTimeNSec);
-
-	long curTimeSec = 0;
-
 	m_CaptureThreadStarted = true;
 	m_StopThread = false;
 
-	if (timeout <= 0)
+	const int64_t timeoutMs = timeout * 1000; // timeout unit is seconds, let's change it to milliseconds
+	auto startTime = std::chrono::steady_clock::now();
+	auto currentTime = startTime;
+
+#if !defined(_WIN32)
+	struct pollfd pcapPollFd;
+	memset(&pcapPollFd, 0, sizeof(pcapPollFd));
+	pcapPollFd.fd = m_PcapSelectableFd;
+	pcapPollFd.events = POLLIN;
+#endif
+
+	if(timeoutMs <= 0)
 	{
 		while (!m_StopThread)
 		{
 			pcap_dispatch(m_PcapDescriptor, -1, onPacketArrivesBlockingMode, (uint8_t*)this);
 		}
-		curTimeSec = startTimeSec + timeout;
 	}
 	else
 	{
-		while (!m_StopThread && curTimeSec <= (startTimeSec + timeout))
+		while (!m_StopThread && std::chrono::duration_cast<std::chrono::milliseconds>(currentTime  - startTime).count() <= timeoutMs )
 		{
-			long curTimeNSec = 0;
-			pcap_dispatch(m_PcapDescriptor, -1, onPacketArrivesBlockingMode, (uint8_t*)this);
-			clockGetTime(curTimeSec, curTimeNSec);
+			if(m_UsePoll)
+			{
+#if !defined(_WIN32)
+				int64_t pollTimeoutMs = timeoutMs - std::chrono::duration_cast<std::chrono::milliseconds>(currentTime  - startTime).count();
+				pollTimeoutMs = std::max(pollTimeoutMs, (int64_t)0); // poll will be in blocking mode if negative value
+				int ready = poll(&pcapPollFd, 1, pollTimeoutMs); // wait the packets until timeout
```

Comment:
I ran the test locally and noticed that `poll()` always returns `1` even if there is no network traffic. Did you test it on your device and see `poll()` timing out? 🤔

Reply:
I don't know how you tested it, did you make sure the `onPacketArrive` callback returns false?

In addition, I think both you and agree that the logic in test file is correct.

please try to run the following code with and without `system()` lines.

```c++
#include <iostream>
#include <pcapplusplus/PcapLiveDevice.h>
#include <pcapplusplus/PcapLiveDeviceList.h>

static bool packetArrivesBlockingModeTimeout(pcpp::RawPacket* rawPacket, pcpp::PcapLiveDevice* dev, void* userCookie)
{
    (*(int*)userCookie)++;
	return false;
}


int main(int argc, char *argv[])
{

	std::system("sudo iptables -A INPUT -i lo -j DROP");
	std::system("sudo iptables -A OUTPUT -o lo -j DROP");

    pcpp::PcapLiveDevice *dev = pcpp::PcapLiveDeviceList::getInstance().getPcapLiveDeviceByIpOrName("lo");


	// open device
	pcpp::PcapLiveDevice::DeviceConfiguration defaultConfig;
	pcpp::PcapLiveDevice::DeviceConfiguration newConfig;
	newConfig.usePoll = true;

	dev->open(newConfig);


	int packetCount = 0;

    /// block 2 second
	int result = dev->startCaptureBlockingMode(packetArrivesBlockingModeTimeout, &packetCount, 2);

    std::cout << "result:" << result << std::endl;
    std::cout << "packets:" << packetCount << std::endl;

	std::system("sudo iptables -D INPUT -i lo -j DROP");
	std::system("sudo iptables -D OUTPUT -o lo -j DROP");

    return 0;
}
```

this is what I get, dropping packets:
```
result:-1
packets:0
```
receiving packets:
```
result:-1
packets:34
```

Reply:
let me investigate a bit

Reply:
Yes, I did run on the interface with incoming packets.

Did you try to use tcpdump or wireshark to monitor the interface and to see if the packets are really blocked? if `poll` returns `1`, it means there are still packets receiving.

a dummy interface could be fine, but I am afraid that it will become too complicated.

Reply:
@seladb I attempted to set up a dummy interface, and while it was successful on Linux, I encountered challenges creating a similar interface on macOS. I propose just using the `eth0` interface for testing purposes. Moreover, if we both agree with the accuracy of the unit test, we can confidently consider the functionality as correct.

Reply:
> I attempted to set up a dummy interface, and while it was successful on Linux, I encountered challenges creating a similar interface on macOS

Yes you're right, this doesn't work on macOS 😕 

> I propose just using the `eth0` interface for testing purposes

If I remember correctly, the interface isn't by default `eth0` on non-Linux OS's (macOS and FreeBSD), and of course people may want to run the tests locally and their default interface won't necessarily be `eth0`. So it's better to do what I suggested here: https://github.com/seladb/PcapPlusPlus/pull/1245#discussion_r1412769322

>  Moreover, if we both agree with the accuracy of the unit test, we can confidently consider the functionality as correct.

I'm not sure `iptables` causes `poll` to time out, I want to double-check that. The problem with the current test is that it passes even if `poll` didn't time out which is not what we want... but I don't know how to improve that and make sure the timeout is because of `poll`

Reply:
**EDIT:** I tested it again, and indeed: when `iptables` drops all packets, `poll` returns `1`. I ran `tcpdump` and Wireshark - both of them don't capture packets related to that interface. They do capture other packets because the (virtual) interface was in promiscuous mode. When I disabled this mode `tcpdump` and Wireshark capture almost no packets, but `poll` still returns `1` all the time...

Reply:
> If I remember correctly, the interface isn't by default eth0 on non-Linux OS's (macOS and FreeBSD),..

Yes, sorry for confusing you, I want to say, use the default interface for doing the test.

> I'm not sure iptables causes poll to time out, I want to double-check that. The problem with the current test is that it passes even if poll didn't time out which is not what we want..

I think I can create a test to prove that it will not timeout without `poll` help. 

>  but poll still returns 1 all the time...

it's intereseting. According to `poll` documents
```
On success, poll() returns a nonnegative value which is the number  of  elements  in  the
       pollfds  whose revents fields have been set to a nonzero value (indicating an event or an
       error).  A return value of zero indicates that the system call timed out before any  file
       descriptors became read.
```

as long as `poll` implementatoin is correct, it shouldn't return `1` if there is no packet.

Since I cannot reproduce your result, could you try to use a different machine or maybe other VMs and see if you can still get the same result?

Reply:
> Since I cannot reproduce your result, could you try to use a different machine or maybe other VMs and see if you can still get the same result?

I'll test it on a different VM and keep you posted

>I think I can create a test to prove that it will not timeout without poll help.

Just curious - what implementation do you have in mind?

Reply:
@seladb  please refer to the new added `TestPcapLiveDeviceBlockingModePollNotTimeout`